llvm/test/CodeGen/AMDGPU/promote-alloca-loadstores.ll

```diff
@@ -198,3 +167,39 @@ entry:
   %tmp = load ptr addrspace(3), ptr addrspace(5) %alloca, align 8
   ret ptr addrspace(3) %tmp
 }
+
+; Will not vectorize because we are doing a load/store of a pointer across
+; address spaces of varying pointer sizes.
+define ptr @alloca_load_store_ptr64_full_ivec(ptr %arg) {
+; CHECK-LABEL: define ptr @alloca_load_store_ptr64_full_ivec
+; CHECK-SAME: (ptr [[ARG:%.*]]) {
+; CHECK-NEXT:  entry:
+; CHECK-NEXT:    [[ALLOCA:%.*]] = alloca [8 x i8], align 8, addrspace(5)
+; CHECK-NEXT:    store ptr [[ARG]], ptr addrspace(5) [[ALLOCA]], align 8
+; CHECK-NEXT:    [[TMP:%.*]] = load ptr, ptr addrspace(5) [[ALLOCA]], align 8
+; CHECK-NEXT:    ret ptr [[TMP]]
+;
+entry:
+  %alloca = alloca [8 x i8], align 8, addrspace(5)
+  store ptr %arg, ptr addrspace(5) %alloca, align 8
+  %tmp = load ptr, ptr addrspace(5) %alloca, align 8
+  ret ptr %tmp
+}
+
+; Will not vectorize because we are doing a load/store of a pointer across
+; address spaces of varying pointer sizes.
```

Comment:
The access sizes are the same. This is still valid, even if the code is doing illegal type punning. The transform should introduce casts to fix the types.

Reply:
Should the amdgpu promote alloca pass do this transformation or should this pass expect the IR to be in a state where pointer storages across address space are done after using an addrspacecast .

Reply:
There's no reason to be talking about addrspacecast. This is just bytes. You do not need to know anything about how they are used or interpreted. The promotion should be done, and it's not important that these are pointer typed

Reply:
Example 1 should be promoted. Example 2 could be promoted, with caveats. It's a UB out of bounds store, so it's not important to handle it optimally. The code merely needs to not crash on this case. In an ordinary optimization pipeline, it would have been deleted already. 

`%alloca = alloca i32, addrspace(5); store ptr %arg, ptr addrspace(5) %alloca` is another out of bounds store that the pass should just avoid crashing on

Reply:
Thanks Matt